### DISCUSSION TITLE
Adding XpressNet

### DISCUSSION THREAD

**Discussion by bask185** [DCC-EX/CommandStation-EX#286] [General]
I am trying to add XpressNet to DCC++ EX. I am using the XpressNetMasterClass of Phillip Gathow. The job itself is not that hard but I would like to know which Serial Port I should use and why. I tried Serial port 1 and 3, but I keep getting interrupt conflicts.
I have some difficulties understanding all of the code since it is quite alot. I am guessing that all 4 are in use, how I can I easily disable one of them? And which port should I use?

**Comment by Asbelos** [comment]
Assuming you have enabled WiFi, the CS code will try to find the wifi chip by sending AT commands to serial1, 2, 3 in turn. 
So if you want to use a serial for your code you must make sure the wifi is on a serial before your device...  otherwise you will be sent an AT command at 112500 baud.

**Comment by Asbelos** [comment]
In addition  you mustensure thattheexpressnet library contains no blocking delays, does not interfere with timer1, does not interfere with any pins used by the motor shields .

Bear in mind this restriction list may be expanded in future dccex versions so if you wish to add expressnet support to our code in the long term it might be necessary to rewrite the library to be a fully configurable and portable part of dcc-ex.

**Comment by habazut** [comment]
I'd encourage you to join our discord and get some help there how to start that project.

You can enable an additional serial port to parse commands by defining for example SERIAL3_COMMANDS. That will enable the port so that incoming messages are sent to the parser selector CommandDistributor.

Then your can look at CommandDistributor.cpp on the devel branch to see how clients are registered and parsed. If you can somehow identify the incoming type you can add a new type and then hopefully hand off the parsing to the library, with a shim class if necessary. If it can not be determined from the first packet if it is expressnet then you need to tie a port to a type but it would be great if the CommandDistributor would work for this. Today both  withrottle and DCCEX protocol parsers are called from there.

Check that the code you take from pghatow is GPL compatible, the licenses found there vary.

Nice to see new coding efforts,
Harald.

**Comment by bask185** [comment]
*****************************************************************************
  *		XpressNetMaster.h - library for XpressNet protocoll
  *		Copyright (c) 08/2016 - 2020 Philipp Gahtow  All right reserved.
  *
*****************************************************************************
Is this an obstacle?

Technical part:
The libary leaves all timers untouched and relies on millis(). It also does not contains blocking loops.
The RS485 direction pin is configurable in the init method. One other configuration are the speed steps. But I default it to 128 steps without problems.

What is a bit of an inconvenience is that the library has it's own instance of a Serial ISR. If I want to use Serial port 0 I cannot type `Serial.` anywhere in the program. The compiler will see the same interrupt being used twice and tosses an error like this one
```
HardwareSerial3.cpp.o (symbol from plugin): In function `Serial3':
(.text+0x0): multiple definition of `__vector_54'
```
Xnet needs to use 9 databits. The default arduino library do not support this AFAIK. Only for sending IIRC

I downloaded discord and made an acount.

I have a side question. I came accross
```
  LCN_SERIAL.begin(115200);
```
I cannot find a definition for `LCN_SERIAL` anywhere in the cloned github repository? What is it?

**Comment by Asbelos** [comment]
Firstly, it is completely unacceptable to interfere with the vendor supplied runtime library as this would compromise all of a users sketches, would need to be edited into multiple device support files for different targets and would make our code un-compilable.

Serial is used throughout dcc-ex and that isn't going to  change. 
However... if all you need is for us to Not reference Serial3 anywhere, then we can easily do that with a define check in wifiinterface.

LCN_SERIAL is something that might be defined in config.h for a specific commercial hook. You can ignore that as the code won't use it if it's not defined.

**Reply by bask185** [reply]
I know it is a huge payne. Every time I had to debug an UNO or NANO using this library, I couldn't make use of the hardware serial port. I used a software serial port to relay debug information to 2nd arduino. I even read information once from leds on an 8 channel relay PCB. Some arduino libs like SD.h have some unused `Serial` remnants which can trigger the compiler bug.

I can refactor the Xnet library to work only if `X_NET` is defined in a config file (and also not-define one of the serial ports in the Wifi part).

The library also needs access to all active locomotives and turnouts in order to issue DCC commands and to respond to all Xnet slaves information requests.

**Comment by Asbelos** [comment]
If it requires a config file, it's no longer suitable to be a "library" so you will need to bring the code into our sketch where it has access to the users config.h values and various cpu type support functions.. and can be omitted entirely if the user does not define the relevant serial etc.

**Comment by Asbelos** [comment]
Bear in mind, from your prev post, you probably already know this...
  DO NOT ISSUE DCC commands for turnouts.
  Do not support 28 speed... this is done automatically at a much lower level. Throttles continue to work 127 only.

**Comment by bask185** [comment]
> If it requires a config file

It does not but I ment the same as what you suggest.
> bring the code into our sketch where it has access to the users config.h values

I simply wanted to add  a line with `// #define X_NET` to the existing user config file. And perhaps also `// #define X_NET_DIR_PIN`

> DO NOT ISSUE DCC commands for turnouts.

Ok! I had not figured that out yet. Does this mean that no Xnet device may issue turnout commands at all or ?

This library comes with callback functions. From void loop() I normally call `Xnet.update()` and from `update()` extern functions are called like:
```
void notifyXNetLocoDrive128( uint16 Address, uint8 Speed )
{
  // do something with loco instruction
}
```
I can stuff all these functions in `XpressNetMasterClass.cpp` and interface somehow to the present loco slot system.

**Comment by habazut** [comment]
Yes, the absence of a license is a problem. (There is only a copyright notice and no license). Things you can NOT do: Copy code from that project into any other project (like DCC-EX) and then publish. As I understand the legal bits: The only thing you might do is to download from pghatow seperately and build something together "in your private corner" and then not share that further.

So I see what is found at pghatow more as an example what can be done. And the code is not that cool that I would want to copy it anyway.

As is we can not copy code from there into DCC-EX.

On the tech side: Is this doing RS485 in software? In that case, would it not be much better to just convert to 8-bit serial with a cheap converter and then use the UART(s) on the Arduino in the usual way?

Harald.

**Comment by Asbelos** [comment]
Turnouts are OK but to throw or close them you must call the relevant functions in the Turnout class. This is because we handle many kinds of turnout hardware at the same time and we have automations that may be listening for turnout changes or broadcasting them to other throttles.

**Reply by bask185** [reply]
> On the tech side: Is this doing RS485 in software? In that case, would it not be much better to just convert to 8-bit serial with a cheap converter and then use the UART(s) on the Arduino in the usual way?

Rs485 is used. Typically with a max485 IC. The usual arduino way is not possible because the  stream library uses 8 bit arrays for it's buffers. Xnet relies on 9 bit. The 9th bit does not fit the array and get's lost.

I mailed mr Gathow if he is is willing to either release his Xnet lib under GPL licence or give explicit permission for usage in this project (provided that is possible). I explained the situation. 

If he is unwilling I will rebuilt a new Xnet library from scratch.  I made a small start on this some months ago.

**Comment by habazut** [comment]
Explicit permission is not enough due to that we don't want to handle DCC-EX as two parts (which would be necessary if the other part is not GPL).

On the RS485 9-bit issue I wonder if it would not be smart to let some dedicated HW handle the filtering (if I understad the 9 bit messages  correct then these are used to identify addresses to make an easier select "this is for me") and then hand over the 8-bit data to DCC-EX when it has been clear that the data is for this device. I don't know much about RS485 but feel that there should be such adapters.

Harald.

**Comment by Asbelos** [comment]
Maybe this helps https://forum.arduino.cc/t/how-to-read-rs485-data-in-arduino/673285

**Reply by bask185** [reply]
No unfortunately not.  But tnx anyways. There is nothing there which is relevant. They just work with 8 bit mode.

**Comment by Ash-4** [comment]
Another example of using RS485 is CMRI.  fwiw

In 2020, I used JMRI/RS485/CMRI for accessories connected to I2C modules.
https://www.trainboard.com/highball/index.php?threads/24-in-48-out-card-for-jmri.116454/page-2#post-1141569

**Reply by bask185** [reply]
I am aware of JMRI, there is nothing there to use. They also work with 8 data bits and have their own protocol.

**Comment by habazut** [comment]
The 9th bit...
https://forum.arduino.cc/t/mods-to-hardwareserial-to-handle-9-bit-data/89447/28
Nick Gammon wrote an alternative lib. Most recent is for IDE 1.6.9. 
My first approach would be a arduino mega that translates 9 bit to 8 bit.

On RX when the code has figured out "this is for me" then there the 9th is not needed any more.
On TX the 9th must be added when sending an addr.
Correct?

Harald.

**Reply by bask185** [reply]
The alternative library replaces the existing hardware library. If you/we start using that, all users of DCC++ EX need to replace the original libraries by the ones of Nick Gammon. His fix exists out of simply changing all 8 bit buffer arrays by 16 bit buffer arrays and adding the actual 9 bit option for configuring the UART.

**Comment by bask185** [comment]
I also do lots of hardware development. I have made an XpressNet PCB with an RJ12 connector (Xnet connector), max485 and a breakout for 5V, GND, Rx and TX. You can plug or solder this PCB with 4 jumper cables in an arduino. If any of you has like a brilliant idea about this, please share it with me so I can make something betters ;)

I have increased my efforts in completely rewriting and rebuiling the library. It will propably take me atleast a month or so. But I will release it under GPL flag :-D It is not a copy paste + reformatting I am truly rewriting the entire thing complete with a different structure. It gives me a better insight in how XpressNet precisely works.

**Comment by bask185** [comment]
query: If an Xnet device issues a command, can I just simply call `DCC::setThrottle()` and `DCC::setAccessory` to ensure the data is transmitted on the DCC bus?

Also If an Xnet device requests information, how do I obtain data from the other files?

Isn't there a short manual for developers for how to interface with the DCC system and sensors n such?

**Reply by bask185** [reply]
I found `DCC::getThrottle()` and similar functions but I am not yet sure in how to to obtain turnout status.

**Comment by Asbelos** [comment]
Turnout::isClosed(id)

**Comment by fodi666** [comment]
Did you manage to get some progress @bask185 ? I'm also interested in such a setup where I could add a LokMaus with an Xpressnet interface to DCC-EX

**Comment by bask185** [comment]
Not yet. I was about half way. I have been more been busy making a line of Open Source DCC decoders. They do lack an english translation though... 
https://train-science.com/OS-ServoDriver/
https://train-science.com/ossd-open-source-solenoid-decoder/

And I am busy with Loconet products. Among others I made a DCC - Lnet command station which is supposed to cost people around €15 (PSU not including) but It need more software work. So far I have tested the accessory commands with succes though.

You can however use the code base of Gathow and make an Xnet interface and use SoftwareSerial to relay the commands in that DCC-EX protocol.  You can use an extra arduino nano or uno for this purpose. I do not know if the lok maus also works well with it? The lok maus, being quite old, seems to give problems on some command stations. Multimaus does work great. I done a bucket load of projects with it. PWM contollers, eavesdroppening devices which can remember routes. And you get to use loco functions to set turnouts and such.

I also have a bare minimum Xnet interface for an arduino.  Look for Xnet.png to see what it is. Gerbers are there too.
https://github.com/bask185/Train-Science-DIY/tree/master/PCB_RELEASES

Is there already an easy to use library to use the DCC ex protocol? Idealy one with software serial possibility. Than I can fix you atleast some code.

**Comment by habazut** [comment]
Everyone writing their own XpressNet parser is not effective. Have a look at how XpressNet over UDP is parsed in the DCC-EX branches that deal with connections from the Z21 app.
Probably the UDP receiver and the parsing should be separated.
Harald.
